test/fail_compilation/retscope6.d

```diff
@@ -21,3 +21,120 @@ int* test() @safe
     arr[0] ~= &i;
     return arr[0][0];
 }
+
+/* TEST_OUTPUT:
+---
+fail_compilation/retscope6.d(7034): Error: reference to local variable `i` assigned to non-scope parameter `_param_1` calling retscope6.S.emplace!(int*).emplace
+fail_compilation/retscope6.d(7035): Error: reference to local variable `i` assigned to non-scope parameter `_param_0` calling retscope6.S.emplace2!(int*).emplace2
+fail_compilation/retscope6.d(7024): Error: scope variable `_param_2` assigned to `s` with longer lifetime
+fail_compilation/retscope6.d(7025): Error: scope variable `_param_2` assigned to `t` with longer lifetime
+fail_compilation/retscope6.d(7037): Error: template instance `retscope6.S.emplace4!(int*)` error instantiating
+---
+*/
+
+#line 7000
+
+alias T = int*;
+
+struct S
+{
+    T payload;
+
+    static void emplace(Args...)(ref S s, Args args) @safe
+    {
+        s.payload = args[0];
+    }
+
+    void emplace2(Args...)(Args args) @safe
+    {
+        payload = args[0];
+    }
+
+    static void emplace3(Args...)(S s, Args args) @safe
+    {
+        s.payload = args[0];
+    }
+
+    static void emplace4(Args...)(scope ref S s, scope out S t, scope Args args) @safe
+    {
+        s.payload = args[0];
+        t.payload = args[0];
+    }
+
+}
+
+void foo() @safe
+{
+    S s;
+    int i;
+    s.emplace(s, &i);
+    s.emplace2(&i);
+    s.emplace3(s, &i);
+    s.emplace4(s, s, &i);
+}
+
+
+/* TEST_OUTPUT:
+---
+fail_compilation/retscope6.d(8016): Error: reference to local variable `i` assigned to non-scope parameter `s` calling retscope6.frank!().frank
+fail_compilation/retscope6.d(8031): Error: reference to local variable `i` assigned to non-scope parameter `p` calling retscope6.betty!().betty
+fail_compilation/retscope6.d(8031): Error: reference to local variable `j` assigned to non-scope parameter `q` calling retscope6.betty!().betty
+fail_compilation/retscope6.d(8048): Error: reference to local variable `j` assigned to non-scope parameter `q` calling retscope6.archie!().archie
+---
+*/
+
+// https://issues.dlang.org/show_bug.cgi?id=19035
+
+#line 8000
+@safe
+{
+
+void escape(int*);
+
+/**********************/
+
+void frank()(ref scope int* p, int* s)
+{
+    p = s;  // should error here
+}
+
+void testfrankly()
+{
+    int* p;
+    int i;
+    frank(p, &i);
+}
+
+/**********************/
+
+void betty()(int* p, int* q)
```

Comment:
It's interesting that you chose to use a template here?  Why is that?  Is it to ensure the body of the function is needed in order to do the inference.  For example, I noticed that Example 1 does not compile, but Example 2 does.

Example 1
```D
@safe:
void betty(int* p, int* q)  // Notice this is not a template
{
     p = q;
}

void testbetty()
{
    int i;
    int j;
    betty(&i, &j); // Error: reference to local variable i assigned to non-scope parameter p calling betty
                   // Error: reference to local variable j assigned to non-scope parameter q calling betty
}
```

Example 2
```D
@safe:
void betty()(int* p, int* q)  // Notice this is a template
{
     p = q;
}

void testbetty()
{
    int i;
    int j;
    betty(&i, &j);
}
```

There's a lot you're not saying here, and I think you're taking a lot for granted expecting us to review the semantics of this PR.